tpgtools/sample.go

```diff
@@ -0,0 +1,153 @@
+package main
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/golang/glog"
+)
+
+// Samples is the array of samples read from the DCL
+type Samples []Sample
+
+// Sample is the object containing sample data from DCL samples
+type Sample struct {
+	// Name is the name of a sample
+	Name *string
+
+	// Description is a short description of the sample
+	Description *string
+
+	// DependencyFileNames contains the filenames of every resource in the sample
+	DependencyFileNames []string `yaml:"dependency"`
+
+	// PrimaryResource is the filename of the sample's primary resource
+	PrimaryResource *string `yaml:"resource"`
+
+	// Substitutions contains every substition in the sample
+	Substitutions []Substitution `yaml:"extra_substitutions"`
+
+	// Dependencies is a list of objects containing metadata for each sample resource
+	Dependencies []Dependency
+}
+
+// Substitution contains metadata that varies for the sample context
+type Substitution struct {
+	// Substitution is the text to be substituted, e.g. topic_name
+	Substitution *string
+
+	// TestValue is the value of the substituted text for tests
+	TestValue *string `yaml:"test_value"`
+
+	// TestValue is the value of the substituted text for docs
+	DocsValue *string `yaml:"docs_value"`
+}
+
+// Dependency contains data that describes a single resource in a sample
+type Dependency struct {
+	// FileName is the name of the file as it appears in testcases.yaml
+	FileName string
+
+	// HCLLocalName is the local name of the HCL block, e.g. "basic" or "default"
+	HCLLocalName string
+
+	// DCLResourceType is the type represented in the DCL, e.g. "ComputeInstance"
+	DCLResourceType string
+
+	// TerraformResourceType is the type represented in Terraform, e.g. "google_compute_instance"
+	TerraformResourceType string
+
+	// HCLBlock is the snippet of HCL config that declares this resource
+	HCLBlock string
+}
+
+// BuildDependency produces a Dependency using a file and filename
+func BuildDependency(fileName string, b []byte) (*Dependency, error) {
+	fileParts := strings.Split(fileName, ".")
+	if len(fileParts) != 4 {
+		return nil, fmt.Errorf("Invalid sample dependency file name: %s", fileName)
+	}
+
+	localName := fileParts[0]
+	dclResourceType := strings.Title(fileParts[1]) + strings.Title(fileParts[2])
+	terraformResourceType := fmt.Sprintf("google_%s_%s", fileParts[1], fileParts[2])
```

Comment:
Is this format guaranteed to be satisfied? What if the resource name is overridden in tpgtools?

Reply:
We should absolutely throw an error if the format is not met in the file name. There isn't another way to get product+resource metadata otherwise.

Yep, should definitely add a name override check here, thanks.